### PR TITLE
Feature/master person index end to end

### PR DESCRIPTION
**Pull request by Abdul-NHS**
## Description
Tickets:
https://nhsd-jira.digital.nhs.uk/browse/GAL-833
https://nhsd-jira.digital.nhs.uk/browse/GAL-451
https://nhsd-jira.digital.nhs.uk/browse/GAL-440
https://nhsd-jira.digital.nhs.uk/browse/GAL-447

Useful links:
https://github.com/NHSDigital/Javascript-mesh-client
https://nhsd-confluence.digital.nhs.uk/pages/viewpage.action?spaceKey=GAL&title=Daily+Demographics+Feed
https://nhsd-confluence.digital.nhs.uk/display/GAL/CaaS+feed+-+Demographics

## Type of changes

- [ ] Refactoring (non-breaking change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ ] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [ ] I have followed the code style of the project
- [ ] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [ ] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: scripts/config/.gitleaks.toml [hunk 1]
```diff
@@ -15,5 +15,18 @@ regexes = [
   '''(127\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}|10\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}|172\.(1[6-9]|2[0-9]|3[0-1])\.[0-9]{1,3}\.[0-9]{1,3}|192\.168\.[0-9]{1,3}\.[0-9]{1,3}|0\.0\.0\.0|255\.255\.255\.255|8\.8\.8\.8|8\.8\.4\.4|208\.67\.222\.222|208\.67\.220\.220)''',
 ]
 
+# [[rules]]
```
Comment (avi312NHS) [inline scripts/config/.gitleaks.toml:18]: a lot of comments here? needed?

File: terraform/modules/iam_galleri_role/iam.tf [hunk 2]
```diff
@@ -315,7 +315,41 @@ resource "aws_iam_policy" "iam_policy_for_get_lsoa_in_range_lambda" {
   })
 }
 
+# Policy required by validateCaasFeedLambda
```
Comment (avi312NHS) [inline terraform/modules/iam_galleri_role/iam.tf:318]: do we need this?

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js [hunk 3]
```diff
@@ -0,0 +1,460 @@
...
+  BatchWriteItemCommand
+} from "@aws-sdk/client-dynamodb";
+import RandExp from 'randexp'
+import { Readable } from "stream";
+import csv from "csv-parser";
+
+const s3 = new S3Client();
+const client = new DynamoDBClient({ region: "eu-west-2", convertEmptyValues: true });
+
+const ENVIRONMENT = process.env.ENVIRONMENT
```
Comment (avi312NHS) [inline terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:19]: we could default environment if none is set in env vars

File: terraform/modules/iam_galleri_role/iam.tf (hunk 2 above)
Comment (Amar-NHS) [reply terraform/modules/iam_galleri_role/iam.tf:318]: We are leaving our respective Lambda Role policy commented out for now as we are using a common Lambda role for most of our Lambda. Plan is to have separate Lambda role for each Lambda later.

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js (hunk 3 above)
Comment (Abdul-NHS) [reply terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:19]: Its required for when deploying outside the default workspace, i.e. uat

File: terraform/modules/iam_galleri_role/iam.tf (hunk 2 above)
Comment (Abdul-NHS) [reply terraform/modules/iam_galleri_role/iam.tf:318]: ^this, there is a policy limit so this is how we are implementing it for the time being

File: scripts/config/.gitleaks.toml (hunk 1 above)
Comment (Abdul-NHS) [reply scripts/config/.gitleaks.toml:18]: Once we have time to implement secret scanning with the correct format we will enable this as it needs some oversight from the platform team. This is with regards to secrets manager, and secret scanner attributing values as 'secrets' which are not the case

Comment (Abdul-NHS) [reply scripts/config/.gitleaks.toml:18]: This is the current template introduced but needs some refinement once we define what rules we want in place

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+  const response = await dbClient.send(command);
+  return response.$metadata.httpStatusCode;
+}
+
+// returns array with each element representing response of batch upload status
+export async function batchWriteRecords(records, chunkSize, dbClient) {
+  console.log(`Number of records to push to db = ${records.length}`)
+  const sendRequest = [];
+  // handle edge case
+  if (records.length === 0) return sendRequest;
```
Comment (avi312NHS) [inline terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:387]: another one is `records.length === 0 || !Array.isArray(records)`

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+      }
+    }
+  }
+}
+
+// returns string to upload to s3
+export const generateCsvString = (header, dataArray) => {
+  return [
+    header,
+    ...dataArray.map((element) => Object.values(element).join(",")),
```
Comment (avi312NHS) [inline terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:458]: better to store this in a const as map returns a new array so shouldn't need to spread in, also for readability

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js (hunk 3 above)
Comment (avi312NHS) [reply terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:19]: so default environment is uat at the moment which is set somewhere?

File: terraform/src/pollMeshMailboxLambda/helper.js [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  } catch (error) {
+    console.log("Failed: ", error);
+    throw error;
+  }
+}
+
+//generator function yields chunkSegment when desired size is reached + header
+export const chunking = function* (itr, size, header) {
+  let chunkSegment = [header];
+  let tempStr = header;
```
Comment (avi312NHS) [inline terraform/src/pollMeshMailboxLambda/helper.js:41]: better variable name would be good `tempStr` would confuse me :D

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js (hunk 3 above)
Comment (Abdul-NHS) [reply terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:19]: No so default is dev, the environment param is configured within the pipeline. For example when deploying you pass in the env.tfvars file which contains the environment you will be setting, which takes these placeholders and configures them to say uat-1, dev-1 etc

File: terraform/src/pollMeshMailboxLambda/helper.js [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      chunkSegment = [header];
+      tempStr = header;
+    }
+  }
+  if (chunkSegment.length) yield tempStr;
+};
+
+//Allows upload of string from chunks to be uploaded to S3
+export async function multipleUpload(chunk, client, environment) {
+  let count = 0;
```
Comment (avi312NHS) [inline terraform/src/pollMeshMailboxLambda/helper.js:57]: this doesn't seem to be used

File: terraform/src/pollMeshMailboxLambda/pollMeshMailboxLambda.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    }
+  } catch (error) {
+    console.error("Error occurred:", error);
+  }
+};
+
+
+//FUNCTIONS
+//Read in MESH data
+async function run() {
```
Comment (avi312NHS) [inline terraform/src/pollMeshMailboxLambda/pollMeshMailboxLambda.js:76]: async function getHealthStatusCode would be better

Comment (Abdul-NHS) [reply terraform/src/pollMeshMailboxLambda/pollMeshMailboxLambda.js]: resolved ty

**Review by Amar-NHS — APPROVED** [review]

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js (hunk 4 above)
Comment (Abdul-NHS) [reply terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:387]: Good spot, but not necessary to implement

File: terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js (hunk 5 above)
Comment (Abdul-NHS) [reply terraform/src/caasFeedAddRecordsLambda/caasFeedAddRecordsLambda.js:458]: Good spot, not necessary to implement atm

**Pull request merged**